pgx/src/heap_tuple/mod.rs

```diff
@@ -0,0 +1,337 @@
+//! Provides a safe interface to Postgres `HeapTuple` objects.
+use crate::{
+    heap_getattr_raw, pg_sys, AllocatedByPostgres, AllocatedByRust, FromDatum, FromDatumResult,
+    IntoDatum, PgBox, PgTupleDesc, TriggerTuple, TryFromDatumError, WhoAllocated,
+};
+use std::num::NonZeroUsize;
+
+/// Describes errors that can occur when trying to create a new [PgHeapTuple].
+#[derive(Debug, Copy, Clone)]
+pub enum PgHeapTupleCreationError {
+    NullTuple,
+    IncorrectAttributeCount,
+}
+
+pub type PgHeapTupleCreationResult<'a, AllocatedBy> =
+    Result<PgHeapTuple<'a, AllocatedBy>, PgHeapTupleCreationError>;
+
+/// A [PgHepTuple] is a lightweight wrapper around Postgres' [pg_sys::HeapTuple] object and a [PgTupleDesc].
+///
+/// In order to access the attributes within a [pg_sys::HeapTuple], the [PgTupleDesc] is required
+/// to describe its structure.
+///
+/// [PgHeapTuple]s can be created from existing (Postgres-provided) [pg_sys:HeapTuple] pointers, from
+/// [pg_sys::TriggerData] pointers, from a composite datum, or created from scratch using raw Datums.
+///
+/// A [PgHeapTuple] can either be considered to be allocated by Postgres or by the Rust runtime. If
+/// allocated by Postgres, it is not mutable until [::into_owned()] is called.
+pub struct PgHeapTuple<'a, AllocatedBy: WhoAllocated<pg_sys::HeapTupleData>> {
+    tuple: PgBox<pg_sys::HeapTupleData, AllocatedBy>,
+    tupdesc: PgTupleDesc<'a>,
+}
+
+impl<'a> PgHeapTuple<'a, AllocatedByPostgres> {
+    /// Creates a new [PgHeapTuple] from a [PgTupleDesc] and a [pg_sys::HeapTuple] pointer.  The
+    /// returned [PgHeapTuple] will be considered by have been allocated by Postgres and is not mutable
+    /// until [::into_owned()] is called.
+    ///
+    /// ## Safety
+    ///
+    /// This function is unsafe as we cannot guarantee that the [pg_sys::HeapTuple] pointer is valid,
+    /// nor can we guaratee that the provided [PgTupleDesc] properly describes the structure of
+    /// the heap tuple.
+    pub unsafe fn from_heap_tuple(tupdesc: PgTupleDesc<'a>, heap_tuple: pg_sys::HeapTuple) -> Self {
+        Self {
+            tuple: PgBox::from_pg(heap_tuple),
+            tupdesc,
+        }
+    }
+
+    /// Creates a new [PgHeapTuple] from one of the two (`Current` or `New`) trigger tuples.  The returned
+    /// [PgHeapTuple] will be considered by have been allocated by Postgres and is not mutable until
+    /// [::into_owned()] is called.  
+    ///
+    /// ## Errors
+    /// - [PgHeapTupleCreationError::NullTuple] if the specified `TriggerTuple` is `null`
+    pub fn from_trigger_data(
+        trigger_data: &'a pg_sys::TriggerData,
+        which_tuple: TriggerTuple,
+    ) -> PgHeapTupleCreationResult<'a, AllocatedByPostgres> {
+        unsafe {
```

Comment:
Is there a reason this isn't an unsafe function, or the unsafe scope isn't smaller?

Reply:
I'll make it unsafe.